catalog/rocks/catalog.go

```diff
@@ -0,0 +1,241 @@
+	Path Path
+	Type DiffType
+}
+
+// StorageNamespace is the URI to the storage location
+type StorageNamespace string
+
+// RepositoryID is an identifier for a repo
+type RepositoryID string
+
+// Path represents a logical path for an entry
+type Path string
+
+// Ref could be a commit ID, a branch name, a Tag
+type Ref string
+
+// TagID represents a named tag pointing at a commit
+type TagID string
+
+// CommitID is a content addressable hash representing a Commit object
+type CommitID string
+
+// BranchID is an identifier for a branch
+type BranchID string
+
+// TreeID represents a snapshot of the tree, referenced by a commit
+type TreeID string
+
+// StagingToken represents a namespace for writes to apply as uncommitted
+type StagingToken string
+
+// CommonPrefix represents a path prefixing one or more Entry objects
+type CommonPrefix string
+
+// Listing represents either an entry or a CommonPrefix
+type Listing struct {
+	CommonPrefix
+	*Entry
+}
+
+// Interfaces
+type Catalog interface {
+	// entries
+	GetEntry(ctx context.Context, repositoryID RepositoryID, ref Ref, path Path) (*Entry, error)
+	SetEntry(ctx context.Context, repositoryID RepositoryID, branchID BranchID, path Path, entry Entry) error
+	DeleteEntry(ctx context.Context, repositoryID RepositoryID, branchID BranchID, path Path) error
+	ListEntries(ctx context.Context, repositoryID RepositoryID, ref Ref, prefix, from, delimiter string, amount int) ([]Listing, bool, error)
+
+	// refs
+	CreateBranch(ctx context.Context, repositoryID RepositoryID, branchID BranchID, ref Ref) (Branch, error)
+	GetBranch(ctx context.Context, repositoryID RepositoryID, branchID BranchID) (Branch, error)
+	Dereference(ctx context.Context, repositoryID RepositoryID, ref Ref) (CommitID, error)
+	Log(ctx context.Context, repositoryID RepositoryID, commitID CommitID, amount int) ([]Commit, bool, error)
+	ListBranches(ctx context.Context, repositoryID RepositoryID, from BranchID, amount int) ([]Branch, bool, error)
+	DeleteBranch(ctx context.Context, repositoryID RepositoryID, branchID BranchID) error
+
+	// commits
+	Commit(ctx context.Context, repositoryID RepositoryID, branchID BranchID, commit Commit) (CommitID, error)
+	Reset(ctx context.Context, repositoryID RepositoryID, branchID BranchID) error
+	Revert(ctx context.Context, repositoryID RepositoryID, branchID BranchID, ref Ref) error
+
+	// diffs and merges
+	Merge(ctx context.Context, repositoryID RepositoryID, from Ref, to BranchID) (CommitID, error)
+	DiffUncommitted(ctx context.Context, repositoryID RepositoryID, branchID BranchID, from Path, amount int) ([]Diff, bool, error)
+	Diff(ctx context.Context, repositoryID RepositoryID, left, right Ref, from Path, amount int) ([]Diff, bool, error)
+}
+
+// internal structures used by Catalog
+type RepositoryIterator interface {
+	First() (RepositoryID, Repository)
+	Next() (RepositoryID, Repository)
+	SeekGE(BranchID) (RepositoryID, Repository)
+	Close()
+}
+
+type EntryIterator interface {
```

Comment:
I strongly prefer the iterators(scanners) we have in the MVCC catalog. I do not think we need to change the interfaces we use because rocksDB use them. Most our iterators are logical. Let us have a talk on this.

Reply:
Paste your suggested interface and we can discuss it.
As for pros and cons - the previous scanner will no longer be available, so we are not adding another variant.
The above interface is similar to how we communicate with pebble so I assume it will just make the code more similar in concept.
If we prefer the scanner interface we also should consider defining a value that captures the key and the value and not returning 3 values.

Reply:
```suggestion
type EntryIterator interface {
	Next() bool
	Value() (*Path, *Entry)
	Err() error
	SeekGE(Path) (bool)
}
```

Reply:
code updated